include/aspect/material_model/drucker_prager.h

```diff
@@ -0,0 +1,248 @@
+/*
+  Copyright (C) 2015 by the authors of the ASPECT code.
+
+  This file is part of ASPECT.
+
+  ASPECT is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2, or (at your option)
+  any later version.
+
+  ASPECT is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file doc/COPYING.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+
+#ifndef __aspect__model_drucker_prager_h
+#define __aspect__model_drucker_prager_h
+
+#include <aspect/material_model/interface.h>
+#include <aspect/simulator_access.h>
+
+namespace aspect
+{
+  namespace MaterialModel
+  {
+    using namespace dealii;
+
+    /**
+     * A material model that consists of globally constant values for all
+     * material parameters except density and viscosity.
+     *
+     * The model is considered incompressible, following the definition
+     * described in Interface::is_compressible.
+     *
+     * The viscosity is computed according to the Drucker Prager frictional
+     * plasticity criterion based on a user-defined Internal angle of friction $\\phi$
+     * and cohesion $C$. In 3D:
+     * $\sigma_y = \frac{6 C \cos(\phi)}{\sqrt(3) (3+\sin(\phi))} +
+     * \frac{2 P \sin(\phi)}{\sqrt(3) (3+\sin(\phi))}$,
+     * where P is the pressure.
+     * See for example Zienkiewicz, O. C., Humpheson, C. and Lewis, R. W. (1975),
+     * G\'{e}otechnique 25, No. 4, 671-689.
+     * With this formulation we circumscribe instead of inscribe the Mohr Coulomb
+     * yield surface.
+     * In 2D the Drucker Prager yield surface is the same
+     * as the Mohr Coulomb surface:
+     * $\sigma_y = P \sin(\phi) + C \cos(\phi)$.
+     * Note that in 2D for $\phi=0$, these criteria
+     * revert to the von Mises criterion (no pressure dependence).
+     * See for example Thieulot, C. (2011), PEPI 188, 47-68.
+     *
+     * Note that we enforce the pressure to be positive to prevent negative
+     * yield strengths and viscosities.
```

Comment:
Can you describe how this enforcement works? The pressure is of course computed somewhere else, so you can't enforce it to be positive, but you can replace it by zero in the computation of sigma_y whenever it is negative. It would be worth describing this here.